examples/catalog/bin/save_screenshots.dart

```diff
@@ -0,0 +1,139 @@
+  UploadError(this.message);
+  final String message;
+  @override
+  String toString() => message;
+}
+
+void logMessage(String s) { print(s); }
+
+class Upload {
+  Upload(this.fromPath, this.largeName, this.smallName);
+
+  static math.Random random;
+  static final String uriAuthority = 'www.googleapis.com';
+  static final String uriPath = 'upload/storage/v1/b/flutter-catalog/o';
+
+  final String fromPath;
+  final String largeName;
+  final String smallName;
+
+  List<int> largeImage;
+  List<int> smallImage;
+  bool largeImageSaved;
+  int retryCount = 0;
+  bool isComplete = false;
+
+  Duration get timeLimit {
+    if (retryCount == 0)
+      return const Duration(milliseconds: 1000);
+    random ??= new math.Random();
+    return new Duration(milliseconds: random.nextInt(1000) + math.pow(2, retryCount) * 1000);
+  }
+
+  Future<bool> save(HttpClient client, String name, List<int> content) async {
+    try {
+      final Uri uri = new Uri.https(uriAuthority, uriPath, <String, String>{
+        'uploadType': 'media',
+        'name': name,
+      });
+      final HttpClientRequest request = await client.postUrl(uri);
+      request
+        ..headers.contentType = new ContentType('image', 'png')
+        ..headers.add('Authorization', 'Bearer $authorizationToken')
+        ..add(content);
+
+      final HttpClientResponse response = await request.close().timeout(timeLimit);
+      if (response.statusCode == HttpStatus.OK) {
+        await response.drain<Null>();
+      } else {
+        logMessage('Request to save "$name" (length ${content.length}) failed, will retry');
+        logMessage(await response.transform(UTF8.decoder).join());
+      }
+      return response.statusCode == HttpStatus.OK;
+    } on TimeoutException catch (_) {
+      logMessage('Request to save "$name" (length ${content.length}) timed out, will retry');
+      return false;
+    }
+  }
+
+  Future<bool> run(HttpClient client) async {
+    assert(!isComplete);
+    if (retryCount > 2)
+      throw new UploadError('upload of "$fromPath" to "$largeName" and "$smallName" failed after 2 retries');
+
+    largeImage ??= await new File(fromPath).readAsBytes();
+    smallImage ??= encodePng(copyResize(decodePng(largeImage), 400));
+
+    if (!largeImageSaved)
+      largeImageSaved = await save(client, largeName, largeImage);
+    isComplete = largeImageSaved && await save(client, smallName, smallImage);
+
+    retryCount += 1;
+    return isComplete;
+  }
+
+  static bool isNotComplete(Upload upload) => !upload.isComplete;
+}
+
+Future<Null> saveScreenshots(List<String> fromPaths, List<String> largeNames, List<String> smallNames) async {
+  assert(fromPaths.length == largeNames.length);
+  assert(fromPaths.length == smallNames.length);
+
+  List<Upload> uploads = new List<Upload>(fromPaths.length);
+  for (int index = 0; index < uploads.length; index += 1)
+    uploads[index] = new Upload(fromPaths[index], largeNames[index], smallNames[index]);
```

Comment:
I did. The (micro) advantage of coding it this way is that uploads is a fixed-length list.